internal/ipc/ipc.go

```diff
@@ -0,0 +1,173 @@
+package ipc
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/ActiveState/cli/internal/errs"
+	"github.com/ActiveState/cli/internal/ipc/internal/flisten"
+	"github.com/ActiveState/cli/internal/ipc/namespace"
+)
+
+var (
+	msgWidth = 64
+	network  = "unix"
+)
+
+type Namespace = namespace.Namespace
+
+type MatchedHandler func(input string) (resp string, isMatched bool)
+
+type IPC struct {
+	n      *Namespace
+	mhs    []MatchedHandler
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     *sync.WaitGroup
+}
+
+func New(n *Namespace, mhs ...MatchedHandler) *IPC {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	ipc := IPC{
+		n:      n,
+		mhs:    make([]MatchedHandler, 0, len(mhs)+2),
+		ctx:    ctx,
+		cancel: cancel,
+		wg:     &sync.WaitGroup{},
+	}
+
+	ipc.mhs = append(ipc.mhs, pingHandler())
+	ipc.mhs = append(ipc.mhs, mhs...)
+	ipc.mhs = append(ipc.mhs, stopHandler(&ipc))
+
+	return &ipc
+}
+
+func (ipc *IPC) ListenAndServe() error {
+	listener, err := flisten.New(ipc.ctx, ipc.n, network)
+	if err != nil {
+		if !errors.Is(err, flisten.ErrInUse) {
+			return errs.Wrap(err, "Cannot construct file listener")
+		}
+
+		ctx, cancel := context.WithTimeout(ipc.ctx, time.Second*3)
+		defer cancel()
+
+		_, pingErr := NewClient(ipc.n).PingServer(ctx)
+		if pingErr == nil {
+			return ErrInUse
+		}
+
+		if !errors.Is(pingErr, flisten.ErrConnRefused) {
+			return errs.Wrap(err, "Cannot connect to existing socket file")
+		}
+
+		listener, err = flisten.NewWithCleanup(ctx, ipc.n, network)
+		if err != nil {
+			return errs.Wrap(err, "Cannot construct file listener after file cleanup")
+		}
+	}
+	defer listener.Close()
+
+	conns := make(chan net.Conn)
+
+	ipc.wg.Add(1)
+	go func() {
+		defer ipc.wg.Done()
+
+		for {
+			if err := accept(ipc.ctx, conns, listener); err != nil {
+				if !errors.Is(err, context.Canceled) {
+					fmt.Fprintln(os.Stderr, fmt.Errorf("listen and server: %w", err)) // TODO: something more useful, log properly, at least
```

Comment:
Seems like a todo worth actioning. Also no todo's in code please. File jira tickets.